File: captchax/sdk/php/src/CaptchaXClient.php
<?php

declare(strict_types=1);

namespace CaptchaX\SDK;

class CaptchaXClient
{
    private CaptchaConfig $config;
    private array $headers;
    private int $retryTimes;
    private float $timeout;

    public function __construct(CaptchaConfig $config)
    {
        $this->config = $config;
        $this->headers = [
            'Content-Type' => 'application/json',
            'Accept' => 'application/json',
        ];
        $this->retryTimes = $config->getRetryTimes();
        $this->timeout = (float)$config->getTimeout() / 1000;

        if ($config->getAppId() !== null) {
            $this->headers['X-App-ID'] = $config->getAppId();
        }
    }

    public static function create(string $baseUrl): self
    {
        return new self(new CaptchaConfig($baseUrl));
    }

    public function setAppId(string $appId): void
    {
        $this->headers['X-App-ID'] = $appId;
    }

    public function setApiVersion(string $version): void
    {
        $this->config = $this->config->withApiVersion($version);
    }

    public function getApiVersion(): string
    {
        return $this->config->getApiVersion();
    }

    private function getApiPrefix(): string
    {
        return '/api/' . $this->config->getApiVersion();
    }

    private function getBaseUrl(): string
    {
        return rtrim($this->config->getBaseUrl(), '/');
    }

    private function requireAppId(): void
    {
        if (!isset($this->headers['X-App-ID'])) {
            throw new CaptchaXException('appId is required for captcha generation', 400, 400);
        }
    }

    private function request(string $method, string $endpoint, ?array $body = null, ?string $deduplicationId = null): array
    {
        $url = $this->getBaseUrl() . $endpoint;
        $lastException = null;

        for ($attempt = 0; $attempt <= $this->retryTimes; $attempt++) {
            $ch = curl_init();

            $headers = $this->headers;
            if ($deduplicationId !== null) {
                $headers['X-Deduplication-ID'] = $deduplicationId;
            }

            $headerList = [];
            foreach ($headers as $key => $value) {
                $headerList[] = "$key: $value";
            }

            curl_setopt_array($ch, [
                CURLOPT_URL => $url,
                CURLOPT_RETURNTRANSFER => true,
                CURLOPT_TIMEOUT => $this->timeout,
                CURLOPT_CONNECTTIMEOUT => $this->timeout,
                CURLOPT_HTTPHEADER => $headerList,
                CURLOPT_SSL_VERIFYPEER => true,
                CURLOPT_SSL_VERIFYHOST => 2,
            ]);

            if ($method === 'POST') {
                curl_setopt($ch, CURLOPT_POST, true);
                if ($body !== null) {
                    curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($body));
                }
            } elseif ($method === 'PUT') {
                curl_setopt($ch, CURLOPT_CUSTOMREQUEST, 'PUT');
                if ($body !== null) {
                    curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($body));
                }
            } elseif ($method === 'DELETE') {
                curl_setopt($ch, CURLOPT_CUSTOMREQUEST, 'DELETE');
            } else {
                curl_setopt($ch, CURLOPT_HTTPGET, true);
            }

            $response = curl_exec($ch);
            $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
            $error = curl_error($ch);
            curl_close($ch);

            if ($response === false) {
                $lastException = new CaptchaXException("cURL error: $error", 0, 0);
                if ($attempt < $this->retryTimes) {
                    usleep((int)(pow(2, $attempt) * 1000000));
                    continue;
                }
                throw $lastException;
            }

            $data = json_decode($response, true);

            if ($httpCode >= 500 && $attempt < $this->retryTimes) {
                $lastException = new CaptchaXException("Server error: $httpCode", $httpCode, $httpCode);
                usleep((int)(pow(2, $attempt) * 1000000));
                continue;
            }

            if ($httpCode >= 400) {
                $message = $data['message'] ?? "HTTP error: $httpCode";
                $code = $data['code'] ?? $httpCode;
                throw new CaptchaXException($message, (int)$code, $httpCode, $data);
            }

            if (isset($data['code']) && $data['code'] !== 200) {
                throw new CaptchaXException(
                    $data['message'] ?? 'Unknown error',
                    (int)$data['code'],
                    $httpCode,
                    $data
                );
            }

            return $data;
        }

        throw new CaptchaXException(
            "Request failed after " . ($this->retryTimes + 1) . " attempts",
            0,
            0,
            $lastException
        );
    }

    private function get(string $endpoint): array
    {
        return $this->request('GET', $endpoint);
    }

    private function post(string $endpoint, ?array $body = null, ?string $deduplicationId = null): array
    {
        return $this->request('POST', $endpoint, $body, $deduplicationId);
    }

    private function put(string $endpoint, ?array $body = null): array
    {
        return $this->request('PUT', $endpoint, $body);
    }

    private function delete(string $endpoint): array
    {
        return $this->request('DELETE', $endpoint);
    }

    public function healthCheck(): HealthStatus
    {
        $data = $this->get('/health');
        return new HealthStatus($data['data'] ?? []);
    }

    public function generateSliderCaptcha(
        ?int $width = null,
        ?int $height = null,
        ?string $clientInfo = null,
        ?string $scenarioId = null
    ): SliderCaptchaResult {
        $this->requireAppId();

        $body = [
            'app_id' => $this->headers['X-App-ID'],
        ];

        if ($width !== null) {
            $body['width'] = $width;
        }
        if ($height !== null) {
            $body['height'] = $height;
        }
        if ($clientInfo !== null) {
            $body['client_info'] = $clientInfo;
        }
        if ($scenarioId !== null) {
            $body['scenario_id'] = $scenarioId;
        }

        $data = $this->post($this->getApiPrefix() . '/captcha/slider', $body);
        return new SliderCaptchaResult($data['data'] ?? []);
    }

    public function verifySliderCaptcha(string $captchaId, int $targetX, ?int $targetY = null): SliderVerifyResult
    {
        $body = [
            'captcha_id' => $captchaId,
            'target_x' => $targetX,
        ];

        if ($targetY !== null) {
            $body['target_y'] = $targetY;
        }

        $data = $this->post($this->getApiPrefix() . '/captcha/slider/verify', $body);
        return new SliderVerifyResult($data['data'] ?? []);
    }

    public function generateClickCaptcha(
        ?int $charCount = null,
        ?string $clientInfo = null,
        ?string $scenarioId = null
    ): ClickCaptchaResult {
        $this->requireAppId();

        $body = [
            'app_id' => $this->headers['X-App-ID'],
        ];

        if ($charCount !== null) {
            $body['char_count'] = $charCount;
        }
        if ($clientInfo !== null) {
            $body['client_info'] = $clientInfo;
        }
        if ($scenarioId !== null) {
            $body['scenario_id'] = $scenarioId;
        }

        $data = $this->post($this->getApiPrefix() . '/captcha/click', $body);
        return new ClickCaptchaResult($data['data'] ?? []);
    }

    public function verifyClickCaptcha(string $captchaId, array $clicks): ClickVerifyResult
    {
        $body = [
            'captcha_id' => $captchaId,
            'clicks' => array_map(
                fn($click) => $click instanceof CharPosition ? $click->toArray() : $click,
                $clicks
            ),
        ];

        $data = $this->post($this->getApiPrefix() . '/captcha/click/verify', $body);
        return new ClickVerifyResult($data['data'] ?? []);
    }

    public function generatePuzzleCaptcha(
        ?int $width = null,
        ?int $height = null,
        ?string $clientInfo = null,
        ?string $scenarioId = null
    ): PuzzleCaptchaResult {
        $this->requireAppId();

        $body = [
            'app_id' => $this->headers['X-App-ID'],
        ];

        if ($width !== null) {
            $body['width'] = $width;
        }
        if ($height !== null) {
            $body['height'] = $height;
        }
        if ($clientInfo !== null) {
            $body['client_info'] = $clientInfo;
        }
        if ($scenarioId !== null) {
            $body['scenario_id'] = $scenarioId;
        }

        $data = $this->post($this->getApiPrefix() . '/captcha/puzzle', $body);
        return new PuzzleCaptchaResult($data['data'] ?? []);
    }

    public function verifyPuzzleCaptcha(string $captchaId, int $targetX, ?int $targetY = null): PuzzleVerifyResult
    {
        $body = [
            'captcha_id' => $captchaId,
            'target_x' => $targetX,
        ];

        if ($targetY !== null) {
            $body['target_y'] = $targetY;
        }

        $data = $this->post($this->getApiPrefix() . '/captcha/puzzle/verify', $body);
        return new PuzzleVerifyResult($data['data'] ?? []);
    }

    public function batchVerify(array $items, ?string $deduplicationId = null): BatchVerifyResponse
    {
        $body = [
            'items' => array_map(
                fn($item) => $item instanceof BatchVerifyItem ? $item->toArray() : $item,
                $items
            ),
        ];

        $data = $this->post($this->getApiPrefix() . '/captcha/batch/verify', $body, $deduplicationId);
        return new BatchVerifyResponse($data['data'] ?? []);
    }

    public function listScenarios(): array
    {
        $data = $this->get($this->getApiPrefix() . '/captcha/scenarios');
        return [
            'scenarios' => array_map(fn($s) => new Scenario($s), $data['data']['scenarios'] ?? []),
            'total' => $data['data']['total'] ?? 0,
        ];
    }

    public function createScenario(string $name, ?string $description = null, ?string $difficulty = null, ?array $config = null): Scenario
    {
        $body = ['name' => $name];

        if ($description !== null) {
            $body['description'] = $description;
        }
        if ($difficulty !== null) {
            $body['difficulty'] = $difficulty;
        }
        if ($config !== null) {
            $body['config'] = $config;
        }

        $data = $this->post($this->getApiPrefix() . '/captcha/scenarios', $body);
        return new Scenario($data['data'] ?? []);
    }

    public function getScenario(string $scenarioId): Scenario
    {
        $data = $this->get($this->getApiPrefix() . '/captcha/scenarios/' . $scenarioId);
        return new Scenario($data['data'] ?? []);
    }

    public function updateScenario(string $scenarioId, array $updates): Scenario
    {
        $data = $this->put($this->getApiPrefix() . '/captcha/scenarios/' . $scenarioId, $updates);
        return new Scenario($data['data'] ?? []);
    }

    public function deleteScenario(string $scenarioId): array
    {
        $data = $this->delete($this->getApiPrefix() . '/captcha/scenarios/' . $scenarioId);
        return $data['data'] ?? [];
    }

    public function registerWebhook(
        string $appId,
        string $url,
        array $events,
        ?string $secret = null,
        ?array $headers = null
    ): Webhook {
        $body = [
            'app_id' => $appId,
            'url' => $url,
            'events' => $events,
        ];

        if ($secret !== null) {
            $body['secret'] = $secret;
        }
        if ($headers !== null) {
            $body['headers'] = $headers;
        }

        $data = $this->post($this->getApiPrefix() . '/captcha/webhook/register', $body);
        return new Webhook($data['data'] ?? []);
    }

    public function listWebhooks(?string $appId = null): array
    {
        $endpoint = $this->getApiPrefix() . '/captcha/webhook';
        if ($appId !== null) {
            $endpoint .= '?app_id=' . urlencode($appId);
        }

        $data = $this->get($endpoint);
        return [
            'webhooks' => array_map(fn($w) => new Webhook($w), $data['data']['webhooks'] ?? []),
            'total' => $data['data']['total'] ?? 0,
        ];
    }

    public function updateWebhook(string $webhookId, array $updates): Webhook
    {
        $data = $this->put($this->getApiPrefix() . '/captcha/webhook/' . $webhookId, $updates);
        return new Webhook($data['data'] ?? []);
    }

    public function unregisterWebhook(string $webhookId): array
    {
        $data = $this->delete($this->getApiPrefix() . '/captcha/webhook/' . $webhookId);
        return $data['data'] ?? [];
    }

    public function createClientInfo(?array $extra = null): string
    {
        $info = [
            'platform' => 'PHP ' . PHP_VERSION,
            'timestamp' => (int)(microtime(true) * 1000),
        ];

        if ($extra !== null) {
            $info = array_merge($info, $extra);
        }

        return json_encode($info);
    }
}

File: captchax/sdk/java/src/main/java/com/captchax/sdk/CaptchaXClient.java
package com.captchax.sdk;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import okhttp3.*;

import java.io.IOException;
import java.util.*;
import java.util.concurrent.TimeUnit;

public class CaptchaXClient {
    private static final MediaType JSON = MediaType.parse("application/json; charset=utf-8");

    private final String baseUrl;
    private String appId;
    private int timeout;
    private int retryTimes;
    private ApiVersion apiVersion;
    private final OkHttpClient httpClient;
    private final Gson gson;
    private final Map<String, String> defaultHeaders;

    public CaptchaXClient(CaptchaConfig config) {
        if (config.getBaseUrl() == null || config.getBaseUrl().isEmpty()) {
            throw new IllegalArgumentException("baseUrl is required");
        }
        this.baseUrl = config.getBaseUrl().replaceAll("/+$", "");
        this.appId = config.getAppId();
        this.timeout = config.getTimeout();
        this.retryTimes = config.getRetryTimes();
        this.apiVersion = config.getApiVersion();

        this.httpClient = new OkHttpClient.Builder()
                .connectTimeout(this.timeout, TimeUnit.MILLISECONDS)
                .readTimeout(this.timeout, TimeUnit.MILLISECONDS)
                .writeTimeout(this.timeout, TimeUnit.MILLISECONDS)
                .retryOnConnectionFailure(true)
                .build();

        this.gson = new GsonBuilder().create();
        this.defaultHeaders = new HashMap<>();
        this.defaultHeaders.put("Content-Type", "application/json");
        this.defaultHeaders.put("Accept", "application/json");

        if (this.appId != null) {
            this.defaultHeaders.put("X-App-ID", this.appId);
        }
    }

    public CaptchaXClient(String baseUrl) {
        this(new CaptchaConfig(baseUrl));
    }

    public CaptchaXClient(String baseUrl, String appId) {
        this(new CaptchaConfig(baseUrl, appId));
    }

    public void setAppId(String appId) {
        this.appId = appId;
        this.defaultHeaders.put("X-App-ID", appId);
    }

    public void setApiVersion(ApiVersion apiVersion) {
        this.apiVersion = apiVersion;
    }

    public ApiVersion getApiVersion() {
        return this.apiVersion;
    }

    private String getApiPrefix() {
        return "/api/" + apiVersion.getValue();
    }

    private String buildUrl(String endpoint) {
        return baseUrl + endpoint;
    }

    private String post(String endpoint, Map<String, Object> body) throws CaptchaXException {
        return post(endpoint, body, null);
    }

    private String post(String endpoint, Map<String, Object> body, String deduplicationId) throws CaptchaXException {
        String url = buildUrl(endpoint);
        Exception lastException = null;

        for (int attempt = 0; attempt <= retryTimes; attempt++) {
            try {
                RequestBody requestBody = RequestBody.create(gson.toJson(body), JSON);

                Headers.Builder headersBuilder = new Headers.Builder();
                for (Map.Entry<String, String> entry : defaultHeaders.entrySet()) {
                    headersBuilder.add(entry.getKey(), entry.getValue());
                }
                if (deduplicationId != null) {
                    headersBuilder.add("X-Deduplication-ID", deduplicationId);
                }

                Request request = new Request.Builder()
                        .url(url)
                        .post(requestBody)
                        .headers(headersBuilder.build())
                        .build();

                try (Response response = httpClient.newCall(request).execute()) {
                    String responseBody = response.body() != null ? response.body().string() : "";

                    if (!response.isSuccessful()) {
                        throw new CaptchaXException(
                                "HTTP error: " + response.code(),
                                response.code(),
                                response.code()
                        );
                    }

                    return responseBody;
                }
            } catch (CaptchaXException e) {
                throw e;
            } catch (IOException e) {
                lastException = e;
                if (attempt < retryTimes) {
                    try {
                        Thread.sleep((long) Math.pow(2, attempt) * 1000);
                    } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                        throw new CaptchaXException("Request interrupted", ie);
                    }
                }
            }
        }

        throw new CaptchaXException("Request failed after " + (retryTimes + 1) + " attempts: " + lastException.getMessage());
    }

    private String get(String endpoint) throws CaptchaXException {
        String url = buildUrl(endpoint);
        Exception lastException = null;

        for (int attempt = 0; attempt <= retryTimes; attempt++) {
            try {
                Headers.Builder headersBuilder = new Headers.Builder();
                for (Map.Entry<String, String> entry : defaultHeaders.entrySet()) {
                    headersBuilder.add(entry.getKey(), entry.getValue());
                }

                Request request = new Request.Builder()
                        .url(url)
                        .get()
                        .headers(headersBuilder.build())
                        .build();

                try (Response response = httpClient.newCall(request).execute()) {
                    String responseBody = response.body() != null ? response.body().string() : "";

                    if (!response.isSuccessful()) {
                        throw new CaptchaXException(
                                "HTTP error: " + response.code(),
                                response.code(),
                                response.code()
                        );
                    }

                    return responseBody;
                }
            } catch (CaptchaXException e) {
                throw e;
            } catch (IOException e) {
                lastException = e;
                if (attempt < retryTimes) {
                    try {
                        Thread.sleep((long) Math.pow(2, attempt) * 1000);
                    } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                        throw new CaptchaXException("Request interrupted", ie);
                    }
                }
            }
        }

        throw new CaptchaXException("Request failed after " + (retryTimes + 1) + " attempts: " + lastException.getMessage());
    }

    private String put(String endpoint, Map<String, Object> body) throws CaptchaXException {
        String url = buildUrl(endpoint);
        Exception lastException = null;

        for (int attempt = 0; attempt <= retryTimes; attempt++) {
            try {
                RequestBody requestBody = RequestBody.create(gson.toJson(body), JSON);

                Headers.Builder headersBuilder = new Headers.Builder();
                for (Map.Entry<String, String> entry : defaultHeaders.entrySet()) {
                    headersBuilder.add(entry.getKey(), entry.getValue());
                }

                Request request = new Request.Builder()
                        .url(url)
                        .put(requestBody)
                        .headers(headersBuilder.build())
                        .build();

                try (Response response = httpClient.newCall(request).execute()) {
                    String responseBody = response.body() != null ? response.body().string() : "";

                    if (!response.isSuccessful()) {
                        throw new CaptchaXException(
                                "HTTP error: " + response.code(),
                                response.code(),
                                response.code()
                        );
                    }

                    return responseBody;
                }
            } catch (CaptchaXException e) {
                throw e;
            } catch (IOException e) {
                lastException = e;
                if (attempt < retryTimes) {
                    try {
                        Thread.sleep((long) Math.pow(2, attempt) * 1000);
                    } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                        throw new CaptchaXException("Request interrupted", ie);
                    }
                }
            }
        }

        throw new CaptchaXException("Request failed after " + (retryTimes + 1) + " attempts: " + lastException.getMessage());
    }

    private String delete(String endpoint) throws CaptchaXException {
        String url = buildUrl(endpoint);
        Exception lastException = null;

        for (int attempt = 0; attempt <= retryTimes; attempt++) {
            try {
                Headers.Builder headersBuilder = new Headers.Builder();
                for (Map.Entry<String, String> entry : defaultHeaders.entrySet()) {
                    headersBuilder.add(entry.getKey(), entry.getValue());
                }

                Request request = new Request.Builder()
                        .url(url)
                        .delete()
                        .headers(headersBuilder.build())
                        .build();

                try (Response response = httpClient.newCall(request).execute()) {
                    String responseBody = response.body() != null ? response.body().string() : "";

                    if (!response.isSuccessful()) {
                        throw new CaptchaXException(
                                "HTTP error: " + response.code(),
                                response.code(),
                                response.code()
                        );
                    }

                    return responseBody;
                }
            } catch (CaptchaXException e) {
                throw e;
            } catch (IOException e) {
                lastException = e;
                if (attempt < retryTimes) {
                    try {
                        Thread.sleep((long) Math.pow(2, attempt) * 1000);
                    } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                        throw new CaptchaXException("Request interrupted", ie);
                    }
                }
            }
        }

        throw new CaptchaXException("Request failed after " + (retryTimes + 1) + " attempts: " + lastException.getMessage());
    }

    private <T> T parseResponse(String responseBody, Class<T> clazz) throws CaptchaXException {
        ApiResponse<?> response = gson.fromJson(responseBody, ApiResponse.class);
        if (response == null || !response.isSuccess()) {
            String message = response != null ? response.getMessage() : "Unknown error";
            int code = response != null ? response.getCode() : 500;
            throw new CaptchaXException(message, code, 500);
        }
        String dataJson = gson.toJson(response.getData());
        return gson.fromJson(dataJson, clazz);
    }

    public HealthStatus healthCheck() throws CaptchaXException {
        String response = get("/health");
        return parseResponse(response, HealthStatus.class);
    }

    public SliderCaptchaResult generateSliderCaptcha() throws CaptchaXException {
        return generateSliderCaptcha(null, null, null, null);
    }

    public SliderCaptchaResult generateSliderCaptcha(Integer width, Integer height) throws CaptchaXException {
        return generateSliderCaptcha(width, height, null, null);
    }

    public SliderCaptchaResult generateSliderCaptcha(Integer width, Integer height, String clientInfo, String scenarioId) throws CaptchaXException {
        if (appId == null) {
            throw new CaptchaXException("appId is required for captcha generation");
        }

        Map<String, Object> body = new HashMap<>();
        body.put("app_id", appId);
        if (width != null) body.put("width", width);
        if (height != null) body.put("height", height);
        if (clientInfo != null) body.put("client_info", clientInfo);
        if (scenarioId != null) body.put("scenario_id", scenarioId);

        String response = post(getApiPrefix() + "/captcha/slider", body);
        return parseResponse(response, SliderCaptchaResult.class);
    }

    public SliderVerifyResult verifySliderCaptcha(String captchaId, int targetX) throws CaptchaXException {
        return verifySliderCaptcha(captchaId, targetX, 0);
    }

    public SliderVerifyResult verifySliderCaptcha(String captchaId, int targetX, Integer targetY) throws CaptchaXException {
        Map<String, Object> body = new HashMap<>();
        body.put("captcha_id", captchaId);
        body.put("target_x", targetX);
        if (targetY != null) body.put("target_y", targetY);

        String response = post(getApiPrefix() + "/captcha/slider/verify", body);
        return parseResponse(response, SliderVerifyResult.class);
    }

    public ClickCaptchaResult generateClickCaptcha() throws CaptchaXException {
        return generateClickCaptcha(null, null, null);
    }

    public ClickCaptchaResult generateClickCaptcha(Integer charCount) throws CaptchaXException {
        return generateClickCaptcha(charCount, null, null);
    }

    public ClickCaptchaResult generateClickCaptcha(Integer charCount, String clientInfo, String scenarioId) throws CaptchaXException {
        if (appId == null) {
            throw new CaptchaXException("appId is required for captcha generation");
        }

        Map<String, Object> body = new HashMap<>();
        body.put("app_id", appId);
        if (charCount != null) body.put("char_count", charCount);
        if (clientInfo != null) body.put("client_info", clientInfo);
        if (scenarioId != null) body.put("scenario_id", scenarioId);

        String response = post(getApiPrefix() + "/captcha/click", body);
        return parseResponse(response, ClickCaptchaResult.class);
    }

    public ClickVerifyResult verifyClickCaptcha(String captchaId, List<CharPosition> clicks) throws CaptchaXException {
        Map<String, Object> body = new HashMap<>();
        body.put("captcha_id", captchaId);
        List<Map<String, Object>> clicksData = new ArrayList<>();
        for (CharPosition click : clicks) {
            Map<String, Object> clickData = new HashMap<>();
            clickData.put("char", click.getCharacter());
            clickData.put("x", click.getX());
            clickData.put("y", click.getY());
            clicksData.add(clickData);
        }
        body.put("clicks", clicksData);

        String response = post(getApiPrefix() + "/captcha/click/verify", body);
        return parseResponse(response, ClickVerifyResult.class);
    }

    public PuzzleCaptchaResult generatePuzzleCaptcha() throws CaptchaXException {
        return generatePuzzleCaptcha(null, null, null, null);
    }

    public PuzzleCaptchaResult generatePuzzleCaptcha(Integer width, Integer height) throws CaptchaXException {
        return generatePuzzleCaptcha(width, height, null, null);
    }

    public PuzzleCaptchaResult generatePuzzleCaptcha(Integer width, Integer height, String clientInfo, String scenarioId) throws CaptchaXException {
        if (appId == null) {
            throw new CaptchaXException("appId is required for captcha generation");
        }

        Map<String, Object> body = new HashMap<>();
        body.put("app_id", appId);
        if (width != null) body.put("width", width);
        if (height != null) body.put("height", height);
        if (clientInfo != null) body.put("client_info", clientInfo);
        if (scenarioId != null) body.put("scenario_id", scenarioId);

        String response = post(getApiPrefix() + "/captcha/puzzle", body);
        return parseResponse(response, PuzzleCaptchaResult.class);
    }

    public PuzzleVerifyResult verifyPuzzleCaptcha(String captchaId, int targetX) throws CaptchaXException {
        return verifyPuzzleCaptcha(captchaId, targetX, 0);
    }

    public PuzzleVerifyResult verifyPuzzleCaptcha(String captchaId, int targetX, Integer targetY) throws CaptchaXException {
        Map<String, Object> body = new HashMap<>();
        body.put("captcha_id", captchaId);
        body.put("target_x", targetX);
        if (targetY != null) body.put("target_y", targetY);

        String response = post(getApiPrefix() + "/captcha/puzzle/verify", body);
        return parseResponse(response, PuzzleVerifyResult.class);
    }

    public BatchVerifyResponse batchVerify(List<BatchVerifyItem> items) throws CaptchaXException {
        return batchVerify(items, null);
    }

    public BatchVerifyResponse batchVerify(List<BatchVerifyItem> items, String deduplicationId) throws CaptchaXException {
        List<Map<String, Object>> itemsData = new ArrayList<>();
        for (BatchVerifyItem item : items) {
            Map<String, Object> itemData = new HashMap<>();
            itemData.put("captcha_id", item.captchaId);
            itemData.put("type", item.type);
            itemData.put("target_x", item.targetX);
            if (item.targetY != null) itemData.put("target_y", item.targetY);
            if (item.clicks != null) {
                List<Map<String, Object>> clicksData = new ArrayList<>();
                for (CharPosition click : item.clicks) {
                    Map<String, Object> clickData = new HashMap<>();
                    clickData.put("char", click.getCharacter());
                    clickData.put("x", click.getX());
                    clickData.put("y", click.getY());
                    clicksData.add(clickData);
                }
                itemData.put("clicks", clicksData);
            }
            itemsData.add(itemData);
        }

        Map<String, Object> body = new HashMap<>();
        body.put("items", itemsData);

        String response = post(getApiPrefix() + "/captcha/batch/verify", body, deduplicationId);
        return parseResponse(response, BatchVerifyResponse.class);
    }

    public Map<String, Object> listScenarios() throws CaptchaXException {
        String response = get(getApiPrefix() + "/captcha/scenarios");
        return parseResponseAsMap(response);
    }

    public Scenario createScenario(String name, String description, String difficulty, Map<String, Object> config) throws CaptchaXException {
        Map<String, Object> body = new HashMap<>();
        body.put("name", name);
        if (description != null) body.put("description", description);
        if (difficulty != null) body.put("difficulty", difficulty);
        if (config != null) body.put("config", config);

        String response = post(getApiPrefix() + "/captcha/scenarios", body);
        return parseResponse(response, Scenario.class);
    }

    public Scenario getScenario(String scenarioId) throws CaptchaXException {
        String response = get(getApiPrefix() + "/captcha/scenarios/" + scenarioId);
        return parseResponse(response, Scenario.class);
    }

    public Scenario updateScenario(String scenarioId, String name, String description, String difficulty, Map<String, Object> config) throws CaptchaXException {
        Map<String, Object> body = new HashMap<>();
        if (name != null) body.put("name", name);
        if (description != null) body.put("description", description);
        if (difficulty != null) body.put("difficulty", difficulty);
        if (config != null) body.put("config", config);

        String response = put(getApiPrefix() + "/captcha/scenarios/" + scenarioId, body);
        return parseResponse(response, Scenario.class);
    }

    public Map<String, Object> deleteScenario(String scenarioId) throws CaptchaXException {
        String response = delete(getApiPrefix() + "/captcha/scenarios/" + scenarioId);
        return parseResponseAsMap(response);
    }

    public Webhook registerWebhook(String appId, String url, List<String> events, String secret, Map<String, String> headers) throws CaptchaXException {
        Map<String, Object> body = new HashMap<>();
        body.put("app_id", appId);
        body.put("url", url);
        body.put("events", events);
        if (secret != null) body.put("secret", secret);
        if (headers != null) body.put("headers", headers);

        String response = post(getApiPrefix() + "/captcha/webhook/register", body);
        return parseResponse(response, Webhook.class);
    }

    public Map<String, Object> listWebhooks(String appId) throws CaptchaXException {
        String endpoint = getApiPrefix() + "/captcha/webhook";
        if (appId != null) {
            endpoint += "?app_id=" + appId;
        }
        String response = get(endpoint);
        return parseResponseAsMap(response);
    }

    public Webhook updateWebhook(String webhookId, String url, String secret, List<String> events, Boolean enabled, Map<String, String> headers) throws CaptchaXException {
        Map<String, Object> body = new HashMap<>();
        if (url != null) body.put("url", url);
        if (secret != null) body.put("secret", secret);
        if (events != null) body.put("events", events);
        if (enabled != null) body.put("enabled", enabled);
        if (headers != null) body.put("headers", headers);

        String response = put(getApiPrefix() + "/captcha/webhook/" + webhookId, body);
        return parseResponse(response, Webhook.class);
    }

    public Map<String, Object> unregisterWebhook(String webhookId) throws CaptchaXException {
        String response = delete(getApiPrefix() + "/captcha/webhook/" + webhookId);
        return parseResponseAsMap(response);
    }

    public String createClientInfo(Map<String, Object> extra) {
        Map<String, Object> info = new HashMap<>();
        info.put("platform", System.getProperty("os.name"));
        info.put("timestamp", System.currentTimeMillis());
        if (extra != null) {
            info.putAll(extra);
        }
        return gson.toJson(info);
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> parseResponseAsMap(String responseBody) throws CaptchaXException {
        return parseResponse(responseBody, Map.class);
    }
}
